system/lib/wasmfs/wasmfs.cpp

```diff
@@ -61,31 +62,103 @@ __wasi_errno_t __wasi_fd_write(__wasi_fd_t fd,
                                const __wasi_ciovec_t* iovs,
                                size_t iovs_len,
                                __wasi_size_t* nwritten) {
+  if (iovs_len < 0) {
+    return __WASI_ERRNO_INVAL;
+  }
+
   auto openFile = FileTable::get()[fd];
 
   if (!openFile) {
     return __WASI_ERRNO_BADF;
   }
 
-  auto file = openFile.locked().getFile()->dynCast<DataFile>();
+  auto lockedOpenFile = openFile.locked();
+  auto* file = lockedOpenFile.getFile()->dynCast<DataFile>();
 
   // If file is nullptr, then the file was not a DataFile.
+  // TODO: change to add support for symlinks.
   if (!file) {
     return __WASI_ERRNO_ISDIR;
   }
 
   auto lockedFile = file->locked();
 
-  __wasi_size_t num = 0;
+  off_t offset = lockedOpenFile.position();
   for (size_t i = 0; i < iovs_len; i++) {
     const uint8_t* buf = iovs[i].buf;
-    __wasi_size_t len = iovs[i].buf_len;
+    size_t len = iovs[i].buf_len;
+
+    // Check if the sum of the buf_len values overflows a uint64_t (file size).
+    if (addWillOverFlow(wasmfs_offset_t(offset), wasmfs_offset_t(len))) {
+      return __WASI_ERRNO_FBIG;
+    }
+
+    // Check if buf_len specifies a positive length buffer but buf is a
+    // null pointer
+    if (!buf && len > 0) {
+      return __WASI_ERRNO_INVAL;
+    }
+
+    lockedFile.write(buf, len, offset);
```

Comment:
Write can return an error code, which we should check.

Reply:
I can keep it a `__wasi_errno_t` for now, but realistically it could also be a bool.

Reply:
It looks like the current implementations only ever return success, but I could imagine that in the future we would want to be able to return various codes like `EAGAIN`, `EWOULDBLOCK`, `EDQUOT`, `EIO`, etc. from different backends. Anyway the current code should probably either handle the return value or assert that it is OK.